### PR TITLE
feat: add cache plugin

### PR DESCRIPTION
**Pull request by pengzhanbo**
### Before submitting the PR, please make sure you do the following 

- [x] Read the [Contributing Guidelines](https://github.com/vuepress/ecosystem/blob/main/CONTRIBUTING.md).
- [x] Provide a description in this PR that addresses **what** the PR is solving. If this PR is going to solve an existing issue, please reference the issue (e.g. `close #123`).

### What is the purpose of this pull request? 

- [ ] Bug fix
- [x] New feature
- [ ] Other

### Description

Vuepress has been criticized for its slow speed. 

This is because during the initialization phase, Vuepress needs to compile all markdown files. Although optimizations have been made in `vuepress/core` using `Promise.all()`, the cost of `createPage()` remains high, with the main expense coming from `markdown render()`. 

In practical use, as we add various markdown-it plugins, including the time-consuming code highlighting plugin, the compilation time remains reasonable when the number of markdown files is small. 

However, with a larger quantity, the time overhead increases significantly. It is unnecessary to recompile all markdown files every time Vuepress is started. For content that has not changed, skipping compilation can greatly improve startup speed.

When starting the development server to write markdown files, if there are multiple code blocks, typically we only update certain blocks within one modification cycle. However, when compiling the markdown file, it is still necessary to recompile all code blocks, which is unnecessary.

In fact, highlighting is also a relatively time-consuming operation, especially when using multiple transformers in `shiki` or even using `twoslash`, the time consumption is quite severe. Therefore, skipping the compilation of unchanged code blocks can greatly reduce the waiting time during hot updates.

This plugin does not solve the startup time consumption during the initial compilation, and it may even slightly increase the time taken for the first compilation. However, subsequent startups will benefit greatly from optimization, with larger projects seeing more noticeable effects.

The current approach is to back up all caches to a single cache file, while also writing them to the runtime memory, which might be more convenient for fast reading but could potentially consume too much memory.

Another option is to generate a separate cache file for each markdown file, using file system reading to achieve this. However, this introduces I/O overhead, and the individual I/O overhead for this part might be greater than the render overhead, possibly requiring additional computation to determine whether a file should use the cache.

### Screenshots

These startup time screenshots are from the `docs` of this repository.

**Unuse cache plugin**

image

**Use cache plugin, first Launch**

image

**Use cache plugin, Secondary Launch**

image

### CODE REVIEW

**Comment by Mister-Hope** [comment]
Any reason why not to import a LRU package?

File: plugins/tools/plugin-cache/src/node/utils.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+export const hash = (data: string): string =>
+  createHash('md5').update(data).digest('hex')
+
+export const readFile = async <T = any>(filepath: string): Promise<T> => {
+  try {
+    const content = await fs.readFile(filepath, 'utf-8')
+    return JSON.parse(content) as T
+  } catch {
+    return {} as T
```
Comment (Mister-Hope) [inline plugins/tools/plugin-cache/src/node/utils.ts:12]: I would prefer null here for this api,.the needed of empty object might be better placed outside this function.

**Comment by Mister-Hope** [comment]
I think we might need the memory usage comparison on some projects.

File: docs/.vuepress/configs/sidebar/en.ts [hunk 2]
```diff
@@ -116,7 +116,12 @@ export const sidebarEn: SidebarConfig = {
     '/plugins/pwa/remove-pwa',
   ],
 
-  '/plugins/tools/': ['google-tag-manager', 'redirect', 'register-components'],
+  '/plugins/tools/': [
+    'google-tag-manager',
+    'redirect',
+    'register-components',
+    'cache',
```
Comment (meteorlxy) [inline docs/.vuepress/configs/sidebar/en.ts:123]: alphabetical ordering is preferred

File: docs/.vuepress/configs/sidebar/zh.ts [hunk 3]
```diff
@@ -120,6 +120,7 @@ export const sidebarZh: SidebarConfig = {
     'google-tag-manager',
     'redirect',
     'register-components',
+    'cache',
```
Comment (meteorlxy) [inline docs/.vuepress/configs/sidebar/zh.ts:123]: same here

File: plugins/tools/plugin-cache/src/node/utils.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import { createHash } from 'node:crypto'
+import { fs, path } from 'vuepress/utils'
+
+export const hash = (data: string): string =>
```
Comment (meteorlxy) [inline plugins/tools/plugin-cache/src/node/utils.ts:4]: Is it possible to make use of the `hash` util from `vuepress/utils`?

File: plugins/tools/plugin-cache/src/node/utils.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  try {
+    const tmp = path.join(cwd, 'tmp')
+    fs.writeFileSync(tmp, '{}', 'utf-8')
+    const start = performance.now()
+    readFileSync(tmp)
+    const end = performance.now()
+    fs.unlinkSync(tmp)
+    return end - start
+  } catch {
+    return 0.15
```
Comment (meteorlxy) [inline plugins/tools/plugin-cache/src/node/utils.ts:45]: What's the magic number for? For a default / fallback value, better to use a constant instead.

File: plugins/tools/plugin-cache/src/node/utils.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
+import { createHash } from 'node:crypto'
+import { fs, path } from 'vuepress/utils'
+
+export const hash = (data: string): string =>
+  createHash('md5').update(data).digest('hex')
+
+export const normalizeFilename = (filename: string): string =>
+  hash(filename).slice(0, 10)
+
+export const readFile = async <T = any>(
```
Comment (meteorlxy) [inline plugins/tools/plugin-cache/src/node/utils.ts:10]: `fs.readJSON` and `fs.readJSONSync`?

File: plugins/tools/plugin-cache/src/node/utils.ts (hunk 5 above)
Comment (pengzhanbo) [reply plugins/tools/plugin-cache/src/node/utils.ts:45]: I run IO tests with 100+ batches on my computer. The time-consuming falls in the range of approximately 0.1ms to 0.2ms, so I roughly chose 0.15 as a fallback option, which may not be used unless there are issues with read or write

File: plugins/tools/plugin-cache/src/node/utils.ts (hunk 4 above)
Comment (pengzhanbo) [reply plugins/tools/plugin-cache/src/node/utils.ts:4]: I just noticed that `vuepress/utils` has `hash`, as long as the hash it generates is unique and stable, it is completely acceptable to use it.

**Comment by Mister-Hope** [comment]
Is this PR ready for review yet?

**Comment by pengzhanbo** [comment]
> Is this PR ready for review yet?

yes

File: docs/plugins/tools/cache.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+By caching the `markdown render` during the initial startup of the VuePress development server, the plugin skips unnecessary `markdown render` on subsequent startups, thus speeding up the startup process.
+
+## Usage
+
+```bash
+npm i -D @vuepress/plugin-cache@next
+```
+
+It is recommended to place `cachePlugin` as the last item in the `plugins` configuration, as this can ensure maximum utilization of caching.
```
Comment (Mister-Hope) [inline docs/plugins/tools/cache.md:15]: I think this should be better as a tip block

File: docs/plugins/tools/cache.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+```ts
+import { cachePlugin } from '@vuepress/plugin-cache'
+
+export default {
+  plugins: [
+    // ... other plugins
+    cachePlugin({
```
Comment (Mister-Hope) [inline docs/plugins/tools/cache.md:23]:
```suggestion
    // ... other plugins
    
    // place it in the end
    cachePlugin({
```

**Pull request merged**